puppet/lib/puppet/util/puppetdb/char_encoding.rb

```diff
@@ -206,6 +220,137 @@ def self.get_byte(str, index)
     end
   end
 
+  # This EncodingAnalyser is supposed to be used by this module only it tries
+  # to break the string into smaller pieces in order to make line-wise diffing
+  # possible.
+  #
+  # In most circumstances the diff should give enough context to understand the
+  # issue better.
+  class EncodingAnalyser
+    class String
+      attr_reader :filename
+
+      def initialize(string, base_filename, type)
+        @string   = string
+        @type     = type
+        @filename = "#{base_filename}-#{@type}"
+        @errors   = []
+      end
+
+      def encoding
+        @string.encoding
+      end
+
+      def size
+        @string.size
+      end
+
+      def write
+        File.open(@filename, 'w') do |file|
+          file.flock(File::LOCK_EX)
+          file.puts split
+          file.chmod(0666)
+          file.flock(File::LOCK_UN)
+        end
+        filename
+      end
+
+      def errors
+        if @errors.any?
+          ["Errors in #{@type} string", @errors].join("\n")
+        else
+          ''
+        end
+      end
+
+      private
```

Comment:
Meh, don't make methods private, makes them harder to test. We use apidocs with yard conventions to mark methods as private: http://www.rubydoc.info/gems/yard/file/docs/Tags.md#private.